src/mcmc/gibbs.jl

```diff
@@ -1,251 +1,528 @@
+        # were not in the domain of the current sampler.
+        vi = merge(vi, context_local.global_varinfo[])
+        # This merges the new values for all the variables sampled by the current sampler.
+        vi = merge(vi, new_vi_local)
+        push!(states, new_state_local)
     end
+    return Transition(model, vi), GibbsState(vi, states)
+end
 
-    # Compute initial states of the local samplers.
-    states = map(samplers) do local_spl
-        # Recompute `vi.logp` if needed.
-        if local_spl.selector.rerun
-            vi = last(
-                DynamicPPL.evaluate!!(
-                    model, vi, DynamicPPL.SamplingContext(rng, local_spl)
-                ),
-            )
-        end
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    model::DynamicPPL.Model,
+    spl::DynamicPPL.Sampler{<:Gibbs},
+    state::GibbsState;
+    kwargs...,
+)
+    vi = varinfo(state)
+    alg = spl.alg
+    varnames = alg.varnames
+    samplers = alg.samplers
+    states = state.states
+    @assert length(samplers) == length(state.states)
+
+    # TODO: move this into a recursive function so we can unroll when reasonable?
+    for index in 1:length(samplers)
+        # Take the inner step.
+        sampler_local = samplers[index]
+        state_local = states[index]
+        varnames_local = _maybevec(varnames[index])
+        vi, new_state_local = gibbs_step_inner(
+            rng, model, varnames_local, sampler_local, state_local, vi; kwargs...
+        )
+        states = Accessors.setindex(states, new_state_local, index)
+    end
+    return Transition(model, vi), GibbsState(vi, states)
+end
 
-        # Compute initial state.
-        _, state = DynamicPPL.initialstep(rng, model, local_spl, vi; kwargs...)
+"""
+    setparams_varinfo!!(model, sampler::Sampler, state, params::AbstractVarInfo)
 
-        # Update `VarInfo` object.
-        vi = gibbs_varinfo(model, local_spl, state)
+A lot like AbstractMCMC.setparams!!, but instead of taking a vector of parameters, takes an
+`AbstractVarInfo` object. Also takes the `sampler` as an argument. By default, falls back to
+`AbstractMCMC.setparams!!(model, state, params[:])`.
 
-        return state
-    end
+`model` is typically a `DynamicPPL.Model`, but can also be e.g. an
+`AbstractMCMC.LogDensityModel`.
+"""
+function setparams_varinfo!!(model, ::Sampler, state, params::AbstractVarInfo)
+    return AbstractMCMC.setparams!!(model, state, params[:])
+end
 
-    # Compute initial transition and state.
-    transition = Transition(model, vi)
-    state = GibbsState(vi, samplers, states)
+function setparams_varinfo!!(
+    model::DynamicPPL.Model, sampler::Sampler{<:MH}, state::VarInfo, params::AbstractVarInfo
+)
+    # The state is already a VarInfo, so we can just return `params`, but first we need to
+    # update its logprob.
+    # TODO(mhauru) Is this the right context to use?
+    return last(DynamicPPL.evaluate!!(model, params, DynamicPPL.leafcontext(model.context)))
```

Comment:
Yep, that's correct. But maybe add a comment as to why it's okay? E.g.
```suggestion
    # NOTE: Using `leafcontext(model.context)` here is a no-op, as it will be concatenated
    # with `model.context` before hitting `model.f`.
    return last(DynamicPPL.evaluate!!(model, params, DynamicPPL.leafcontext(model.context)))
```